File: LSQFit.py
import numpy as np
from numpy.linalg import inv
from matplotlib import pyplot as plt

xmin=1.0
xmax=20.0
npoints=12
sigma=0.2
lx=np.zeros(npoints)
ly=np.zeros(npoints)
ley=np.zeros(npoints)
pars=[0.5,1.3,0.5]

from math import log
def f(x,par):
    return par[0]+par[1]*np.log(x)+par[2]*np.log(x)*np.log(x)

from random import gauss
def getX(x):  # x = array-like
    step=(xmax-xmin)/npoints
    for i in range(npoints):
        x[i]=xmin+i*step
        
def getY(x,y,ey):  # x,y,ey = array-like
    for i in range(npoints):
        y[i]=f(x[i],pars)+gauss(0,sigma)
        ey[i]=sigma

# # get a random sampling of the (x,y) data points, rerun to generate different data sets for the plot below
# getX(lx)
# getY(lx,ly,ley)

# fig, ax = plt.subplots()
# ax.errorbar(lx, ly, yerr=ley)
# ax.set_title("Pseudoexperiment")
# fig.show()


# *** modify and add your code here ***
nexperiments = int(1e6)  # for example

# we want to implement linear fit matrix
# define X = [1, ln x_1, ln(x_1)^2], p = [a,b,c]^T 
# y = Xp (model), 
# weight matrix is W_ii = 1/sigma_i^2 (diagonal)
# we minimize (X^T W X)p = X^T W y  

def xmatrix(x): 
    lx = np.log(x)
    return np.column_stack([np.ones_like(lx), lx, lx**2]) 

def wls_fit(x, y, ey): 
    X = xmatrix(x) # (n, 3)
    w = 1 / (ey * ey) # (n, ) 

    # normal equations, 
    XT_W = (X.T * w) # (3, n)
    A = XT_W @ X # (3, 3) # (matrix multiplication ==> @)
    b = XT_W @ y # (3, )
    p = np.linalg.solve(A, b) # [a, b, c]
    
    yhat = X @ p 
    chi2 = np.sum(((y-yhat)/ey)**2)
    dof = len(y) - X.shape[1]
    return p, chi2, dof


rng = np.random.default_rng(12345) 

par_a = np.random.rand(nexperiments)   
par_b = np.random.rand(nexperiments)   
par_c = np.random.rand(nexperiments)
chi2_reduced = np.random.rand(nexperiments)

x = np.empty(npoints, dtype=float)
getX(x)

for i in range(nexperiments):
    y = np.empty(npoints, dtype=float)
    ey = np.empty(npoints, dtype=float) 


    # generate data
    y[:] = f(x, pars) + rng.normal(0.0, sigma, size=npoints) 
    ey[:] = sigma

    p, chi2, dof = wls_fit(x, y, ey)
    par_a[i], par_b[i], par_c[i] = p
    chi2_reduced[i] = chi2 / dof


x = np.empty(npoints)
y = np.empty(npoints)
ey = np.empty(npoints)

getX(x)
getY(x, y, ey)

p, chi2, dof = wls_fit(x, y, ey)
print(f"Fitted parameters: a={p[0]:.4f}, b={p[1]:.4f}, c={p[2]:.4f}")
print(f"$\chi^2$/ndf = {chi2/dof:.3f}")

# make smooth curve for plotting
xs = np.linspace(xmin, xmax, 400)
yfit = f(xs, p)
ytrue = f(xs, pars)

fig, ax = plt.subplots(figsize=(6,4))
ax.errorbar(x, y, yerr=ey, fmt='o', label='Pseudo-data')
ax.plot(xs, yfit, label='WLS fit')
ax.plot(xs, ytrue, '--', label='Truth')
ax.set_xlabel('x')
ax.set_ylabel('y')
ax.set_title(f"Least-Squares Fit ($\chi^2$/ndf = {chi2/dof:.2f})")
ax.legend()
fig.savefig('./LSQFit_py_example.pdf')

fig, axs = plt.subplots(2, 2, figsize=(10,10))
plt.tight_layout()

# careful, the automated binning may not be optimal for displaying your results!
h1 = axs[0, 0].hist2d(par_a, par_b, bins=60,)
axs[0, 0].set_xlabel('a'); axs[0, 0].set_ylabel('b'); axs[0, 0].set_title('Parameter b vs a')
fig.colorbar(h1[3], ax=axs[0,0])

h2 = axs[0, 1].hist2d(par_a, par_c, bins=60,)
axs[0, 1].set_xlabel('a'); axs[0, 1].set_ylabel('c'); axs[0, 1].set_title('Parameter c vs a')
fig.colorbar(h2[3], ax=axs[0,1])

h3 = axs[1, 0].hist2d(par_b, par_c, bins=60,)
axs[1, 0].set_xlabel('b'); axs[1, 0].set_ylabel('c'); axs[1, 0].set_title('Parameter c vs b')
fig.colorbar(h3[3], ax=axs[1,0])

axs[1, 1].hist(chi2_reduced, bins=80, density=True)
axs[1, 1].set_xlabel('$\chi^2$/ndf'); axs[1, 1].set_title('Reduced $\chi^2$ distribution')
plt.tight_layout()
# fig.savefig('./LSQFit.pdf')

from matplotlib.backends.backend_pdf import PdfPages

# Create multi-page PDF
with PdfPages("LSQFit_py.pdf") as pdf:
    # First figure: pseudoexperiment
    fig1, ax1 = plt.subplots(figsize=(6,4))
    ax1.errorbar(lx, ly, yerr=ley, fmt='o', label='Pseudo-data')
    xs = np.linspace(xmin, xmax, 400)
    p, chi2, dof = wls_fit(lx, ly, ley)
    ax1.plot(xs, f(xs, p), label='WLS fit')
    ax1.plot(xs, f(xs, pars), '--', label='Truth')
    ax1.set_title(f"Pseudoexperiment (χ²/ndf = {chi2/dof:.2f})")
    ax1.set_xlabel('x'); ax1.set_ylabel('y')
    ax1.legend()
    pdf.savefig(fig1)        # save this figure as page 1
    plt.close(fig1)

    # Second figure: parameter study (the 2×2 grid)
    fig2, axs = plt.subplots(2, 2, figsize=(9, 8))
    plt.tight_layout(pad=2.0)
    axs[0,0].hist2d(par_a, par_b, bins=60, range=[(0.2,0.8),(0.7,1.9)])
    axs[0,0].set_title('Parameter b vs a')
    axs[0,1].hist2d(par_a, par_c, bins=60, range=[(0.2,0.8),(0.2,0.9)])
    axs[0,1].set_title('Parameter c vs a')
    axs[1,0].hist2d(par_b, par_c, bins=60, range=[(0.7,1.9),(0.2,0.9)])
    axs[1,0].set_title('Parameter c vs b')
    axs[1,1].hist(chi2_reduced, bins=80, density=True)
    axs[1,1].set_title('Reduced χ² distribution')
    for ax in axs.flat:
        ax.grid(True, alpha=0.25)
    fig2.suptitle('Least-squares study results')
    pdf.savefig(fig2)        # save as page 2
    plt.close(fig2)

# **************************************

File: LSQFit.cpp
#include "TRandom2.h"
#include "TGraphErrors.h"
#include "TMath.h"
#include "TApplication.h"
#include "TCanvas.h"
#include "TH2F.h"
#include "TH1F.h"
#include "TGClient.h"
#include "TStyle.h"


#include "TMatrixD.h"
#include "TVectorD.h"
#include "TDecompLU.h"
#include "TGraph.h"
#include "TLegend.h"

#include <vector>
#include <cmath>


#include <iostream>
using namespace std;

using TMath::Log;

//parms
const double xmin=1;
const double xmax=20;
const int npoints=12;
const double sigma=0.2;

double f(double x){
  const double a=0.5;
  const double b=1.3;
  const double c=0.5;
  return a+b*Log(x)+c*Log(x)*Log(x);
}

void getX(double *x){
  double step=(xmax-xmin)/npoints;
  for (int i=0; i<npoints; i++){
    x[i]=xmin+i*step;
  }
}

void getY(const double *x, double *y, double *ey){
  static TRandom2 tr(0);
  for (int i=0; i<npoints; i++){
    y[i]=f(x[i])+tr.Gaus(0,sigma);
    ey[i]=sigma;
  }
}


void leastsq(){
  double x[npoints];
  double y[npoints];
  double ey[npoints];
  getX(x);
  getY(x,y,ey);
  auto tg = new TGraphErrors(npoints,x,y,0,ey);
  tg->Draw("alp");
}

double f_model(double x, double a, double b, double c){
  const double lx = Log(x);   
  return a + b*lx + c*lx*lx;
}

struct FitResult {
  double a,b,c;
  double chi2;
  int dof; 
};

FitResult wls_fit(const double* x, const double* y, const double* ey, int n) {
  // define X matrix 
  TMatrixD X(n, 3);
  TVectorD Y(n), W(n); 

  for (int i = 0; i < n; ++i) {
    const double lx = Log(x[i]);
    X(i, 0) = 1;
    X(i, 1) = lx; 
    X(i, 2) = lx*lx; 
    Y[i] = y[i]; 
    W[i] = 1/(ey[i]*ey[i]);
  }

  // define normal equation 
  TMatrixD XT(3, n); XT.Transpose(X);

    // XT_W = XT * W (W is diagonal → scale each column by W[i])
  TMatrixD XT_W(3, n);
  for (int r=0; r<3; ++r)
    for (int c=0; c<n; ++c)
      XT_W(r,c) = XT(r,c) * W[c];

  TMatrixD A(3,3); A.Mult(XT_W, X);     // A = X^T W X
  TVectorD b(3);   b = XT_W * Y;        // b = X^T W y

  // Solve for parameters p
  TDecompLU lu(A);
  TVectorD p = b;
  if (!lu.Solve(p)) {
    std::cerr << "wls_fit: linear solve failed\n";
    return FitResult{0,0,0,0.0, n-3};
  }

  // Compute chi^2
  double chi2 = 0.0;
  for (int i = 0; i < n; ++i){
    const double lx   = X(i,1);
    const double lx2  = X(i,2);
    const double yhat = p[0] + p[1]*lx + p[2]*lx2;
    const double pull = (Y[i] - yhat) / std::sqrt(1.0/W[i]); // = (Y - yhat)/ey
    chi2 += pull*pull;
  }

  return { p[0], p[1], p[2], chi2, n - 3 };
}

int main(int argc, char **argv){
  TApplication theApp("App", &argc, argv); // init ROOT App for displays

  // ******************************************************************************
  // ** this block is useful for supporting both high and std resolution screens **
  UInt_t dh = gClient->GetDisplayHeight()/2;   // fix plot to 1/2 screen height  
  //UInt_t dw = gClient->GetDisplayWidth();
  UInt_t dw = 1.1*dh;
  // ******************************************************************************

  gStyle->SetOptStat(0); // turn off histogram stats box


  TCanvas *tc = new TCanvas("c1","Sample dataset",dw,dh);

  double lx[npoints];
  double ly[npoints];
  double ley[npoints];

  getX(lx);
  getY(lx,ly,ley);
  auto tgl = new TGraphErrors(npoints,lx,ly,0,ley);
  tgl->SetTitle("Pseudoexperiment;x;y");
  
  // An example of one pseudo experiment
  tgl->Draw("alp");
  tc->Draw();

  
  // *** modify and add your code here ***
  FitResult demo = wls_fit(lx, ly, ley, npoints);

  const int Ncurv = 400;
  std::vector<double> xs(Ncurv), yfit(Ncurv), ytru(Ncurv);
  for (int i=0; i<Ncurv; ++i){
    xs[i] = xmin + (xmax - xmin) * (double)i / (Ncurv-1);
    yfit[i] = f_model(xs[i], demo.a, demo.b, demo.c);
    ytru[i] = f_model(xs[i], /*a_true*/0.5, /*b_true*/1.3, /*c_true*/0.5);
  }
  auto grFit = new TGraph(Ncurv, xs.data(), yfit.data()); grFit->SetLineWidth(2);
  auto grTru = new TGraph(Ncurv, xs.data(), ytru.data()); grTru->SetLineStyle(2);

  tgl->SetTitle(Form("Pseudoexperiment; x; y  (chi2/ndf = %.2f)", demo.chi2/(double)demo.dof));
  grFit->Draw("L SAME");
  grTru->Draw("L SAME");
  auto leg = new TLegend(0.60,0.75,0.88,0.90);
  leg->AddEntry(grFit, "WLS fit","l");
  leg->AddEntry(grTru, "Truth",  "l");
  leg->Draw();
  tc->Modified(); tc->Update();

  // ---------- Study over many pseudo-experiments ----------
  const int nexperiments = (int) 1e6;

  // Choose display ranges near the truth (improves the 2D density look)
  TH2F *h1 = new TH2F("h1","Parameter b vs a; a; b", 80, 0.2, 0.8, 80, 0.7, 1.9);
  TH2F *h2 = new TH2F("h2","Parameter c vs a; a; c", 80, 0.2, 0.8, 80, 0.2, 0.9);
  TH2F *h3 = new TH2F("h3","Parameter c vs b; b; c", 80, 0.7, 1.9, 80, 0.2, 0.9);
  TH1F *h4 = new TH1F("h4","Reduced #chi^{2};;Density", 100, 0.0, 2.5);
  h4->Sumw2();

  // Reuse x positions but regenerate y each time
  double xbuf[npoints];
  getX(xbuf);

  for (int ie = 0; ie < nexperiments; ++ie){
    double ybuf[npoints], eybuf[npoints];
    getY(xbuf, ybuf, eybuf);

    FitResult r = wls_fit(xbuf, ybuf, eybuf, npoints);

    h1->Fill(r.a, r.b);
    h2->Fill(r.a, r.c);
    h3->Fill(r.b, r.c);
    h4->Fill(r.chi2 / r.dof, 1.0 / nexperiments); // scale so it looks like a PDF
  }

  // ---------- Draw results ----------
  TCanvas *tc2 = new TCanvas("c2","my study results",200,200,dw,dh);
  tc2->Divide(2,2);
  tc2->cd(1); gPad->SetRightMargin(0.14); h1->Draw("COLZ");
  tc2->cd(2); gPad->SetRightMargin(0.14); h2->Draw("COLZ");
  tc2->cd(3); gPad->SetRightMargin(0.14); h3->Draw("COLZ");
  tc2->cd(4); h4->Draw("HIST");
  tc2->Modified(); tc2->Update();

  TString outname = "LSQFit_cpp.pdf";

  // open PDF file (first page)
  tc->Print(outname + "(", "pdf");   // "(" means start multipage PDF
  tc2->Print(outname + ")", "pdf");  // ")" closes it

  cout << "Saved results to " << outname << endl;

  // **************************************
  
  // cout << "Press ^c to exit" << endl;
  // theApp.SetIdleTimer(30,".q");  // set up a failsafe timer to end the program  
  // theApp.Run();
}
